backend/tournesol/suggestions/graph.py

```diff
@@ -0,0 +1,310 @@
+from __future__ import annotations
+
+from typing import Optional
+
+import numpy as np
+from django.db.models import Avg, QuerySet
+
+from tournesol.models import (
+    ContributorRatingCriteriaScore,
+    ContributorScaling,
+    EntityCriteriaScore,
+    Poll,
+)
+from tournesol.suggestions.suggested_user import SuggestedUser
+from tournesol.suggestions.suggested_video import SuggestedVideo
+
+
+class Graph:
+    """
+    Class representing a comparison graph
+    Each node is a video and each node is a comparison
+    """
+    # not necessarily the best data structure,
+    # I still have to find out what is the best one between the two
+    edges: list[tuple[SuggestedVideo, SuggestedVideo]]
+    _nodes: list[SuggestedVideo]  # todo clean that, use default dict functions
+    graph: dict[SuggestedVideo, list[SuggestedVideo]]
+    sigma: float
+    dirty: bool
+    _local_user: SuggestedUser
+    local_user_scaling: ContributorScaling
+    _local_poll: Poll
+
+    LAMBDA_THRESHOLD = 0.5
+    MIN_SCALING_ACCURACY = 0.5
+
+    ABSENT_NODE = SuggestedVideo(None)
+
+    adjacency_matrix: np.array
+    normalized_adjacency_matrix: np.array
+
+    distance_matrix: dict[SuggestedVideo, dict[SuggestedVideo, int]]
+    similarity_matrix: np.array
+
+    @property
+    def nodes(self) -> list[SuggestedVideo]:
+        return self._nodes
+
+    def __init__(self, local_user: Optional[SuggestedUser], local_poll: Poll, local_criteria):
+        self.local_user_mean: float = 0
+        self.video_comparison_reference = SuggestedVideo(None)
+        self.dirty = True
```

Comment:
The dirty means here that the information has changed since the last offline computation, and when the class is built, the computed elements are not yet computed so it makes sense to initialize it to True while the offline computation is not done (allowing for example to avoid sending a response that wouldn't make sense if a request is received while the first offline computation is in progress)
The term dirty is often used for that in low level programming, so I thougth it was clear, but it may make sense to rename it ?